### DISCUSSION TITLE
How to differentiate between variables and functions

### DISCUSSION THREAD

**Discussion by tannakartikey** [josdejong/mathjs#2657] [Q&A] [answered]
I want to extract all the variables from a formula.

I am traversing through the Expression Tree using the `traverse` method to get all the _SymbolNodes_.

The problem is that the functions like `sqrt`, `pi`, `sin` are considered as _SymbolNode_.

Is there any way I can ignore all those values? Right now I am doing the following to differentiate them:
```
typeof math[node.name] === 'function' ? 'This is function' : 'This is symbol'
```
The `pi` is not a function either.

I want to do this so that I can ask the user for those values and then set the scope using the user input to evaluate the formula.

Is there any better way to handle this?

**Comment by josdejong** [comment, marked as answer]
Good question. An expression with a function like `sin(2)` is parsed into a `FunctionNode` with a property `.fn` containing a `SymbolNode` with name `sin`. To check whether a `SymbolNode` is the name of a function, you can check whether it's parent is a `FunctionNode` or not.

**Comment by zagushka** [comment]
@tannakartikey 
I'm working on the same task, in addition to @josdejong comment
`sin(x)` parsed to two `SymbolNodes` -  `sin` and `x` with the same `FunctionNode` parent. So additional check parent have the same name is required.
My final check node is variable `node.isSymbolNode && parent?.name !== node.name;`
